frontend/src/metabase-lib/v1/expressions/resolver.unit.spec.ts

```diff
@@ -1,438 +1,468 @@
-      expect(filter(`contains([C], "somestring")`).fields).toEqual(["C"]);
-      expect(filter(`doesNotContain([C], "somestring")`).fields).toEqual(["C"]);
-    });
+      it("should resolve segments", () => {
+        expect(boolean("Expensive Things")).toEqual(segments.EXPENSIVE_THINGS);
+      });
 
-    it("should work on functions with optional flag", () => {
-      expect(() =>
-        filter(`interval([A], 3, "day", "include-current")`),
-      ).not.toThrow();
-    });
-  });
+      it("should resolve columns of type: boolean", () => {
+        expect(boolean("bool")).toEqual(expressions.BOOL);
+      });
 
-  describe("for expressions (for custom columns)", () => {
-    it("should resolve segments correctly", () => {
-      expect(expression(`trim([A])`).segments).toEqual([]);
-      expect(expression(`round([B])`).segments).toEqual([]);
-      expect(expression(`concat([S])`).segments).toEqual([]);
-      expect(expression(`concat([A], [B])`).segments).toEqual([]);
-      expect(expression(`coalesce([P])`).segments).toEqual([]);
-      expect(expression(`coalesce([P], [Q], [R])`).segments).toEqual([]);
-      expect(expression(`notNull([A])`).segments).toEqual([]);
-      expect(expression(`notEmpty([A])`).segments).toEqual([]);
-    });
+      it("should not resolve unknown fields", () => {
+        expect(() => boolean("Unknown")).toThrow(
+          "Unknown Segment or boolean Field: Unknown",
+        );
+      });
 
-    it("should resolve fields correctly", () => {
-      expect(expression(`trim([A])`).fields).toEqual([`A`]);
-      expect(expression(`round([B])`).fields).toEqual(["B"]);
-      expect(expression(`concat([S])`).fields).toEqual(["S"]);
-      expect(expression(`concat([A], [B])`).fields).toEqual(["A", "B"]);
-      expect(expression(`coalesce([P])`).fields).toEqual(["P"]);
-      expect(expression(`coalesce([P], [Q], [R])`).fields).toEqual([
-        "P",
-        "Q",
-        "R",
-      ]);
-      expect(expression(`in([A], [B], [C])`).fields).toEqual(["A", "B", "C"]);
-      expect(expression(`text([A])`).fields).toEqual(["A"]);
-      expect(expression(`integer([A])`).fields).toEqual(["A"]);
-      expect(expression(`doesNotContain([A], "SomeString")`).fields).toEqual([
-        "A",
-      ]);
-      expect(expression(`notNull([A])`).fields).toEqual(["A"]);
-      expect(expression(`notEmpty([A])`).fields).toEqual(["A"]);
-    });
+      it("should not resolve fields with non-boolean types", () => {
+        expect(() => boolean("stringly")).toThrow(
+          "Unknown Segment or boolean Field: stringly",
+        );
+        expect(() => boolean("Total")).toThrow(
+          "Unknown Segment or boolean Field: Total",
+        );
+      });
 
-    it("should allow nested datetime expressions", () => {
-      expect(() => expression(`year(now)`)).not.toThrow();
+      it("should not resolve metrics", () => {
+        // TODO: should this resolve metrics?
```

Comment:
I think so. It's possible to create a metric that will return a boolean (e.g. Accounts > Summarize by Maximum of Trial Converted).

Reply:
@romeovs shouldn't we change how it works?